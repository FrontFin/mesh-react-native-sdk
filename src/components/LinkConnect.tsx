import { View } from 'react-native';
import { WebView } from 'react-native-webview';
import React, { useMemo, useRef, useState } from 'react';

import { NavBar } from './NavBar';
import { SDKContainer } from './SDKContainer';
import { SDKViewContainer } from './SDKViewContainer';

import type { LinkConfiguration } from '../';
import { useSDKCallbacks } from '../hooks/useSDKCallbacks';
import { sdkSpecs } from '../utils/sdkConfig';
import { DARK_THEME_COLOR_BOTTOM, LIGHT_THEME_COLOR_BOTTOM } from '../constant';

export const LinkConnect = (props: LinkConfiguration) => {
  const {
    showNativeNavbar,
    showWebView,
    linkUrl,
    darkTheme,
    handleMessage,
    handleNavState,
    showCloseAlert,
  } = useSDKCallbacks(props);
  const webViewRef = useRef<WebView>(null);
  const goBack = () => webViewRef?.current?.goBack();

  const injectedScript = useMemo(() => {
    let sdkTypeScript = `
    window.meshSdkPlatform='${sdkSpecs.platform}';
    window.meshSdkVersion='${sdkSpecs.version}';
  `;

    if (props.settings) {
      sdkTypeScript += `
        window.accessTokens='${JSON.stringify(
          props.settings.accessTokens || {}
        )}';
        window.transferDestinationTokens='${JSON.stringify(
          props.settings.transferDestinationTokens || {}
        )}';
      `;
    }

    return sdkTypeScript;
  }, [props.settings]);

  const SDKWrapperComponent = props.renderViewContainer
    ? SDKViewContainer
    : SDKContainer;

  const LoadingComponentWebview = () => {
    return (
      <View style={{
        position: 'absolute',
        zIndex: 10,
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        backgroundColor: darkTheme ? DARK_THEME_COLOR_BOTTOM : LIGHT_THEME_COLOR_BOTTOM
      }}/>
    );
  };

  const [initialLoading, setInitialLoading] = useState(true);

  if (darkTheme === undefined) {
    return null
  }
  
  return (
    <SDKWrapperComponent isDarkTheme={darkTheme}>
      {showNativeNavbar && (
        <NavBar goBack={goBack} showCloseAlert={showCloseAlert} isDarkTheme={darkTheme} />
      )}
      {initialLoading && <LoadingComponentWebview />}
      {showWebView && linkUrl && (
        <WebView
          bounces={false}
          style={{ backgroundColor: darkTheme ? DARK_THEME_COLOR_BOTTOM : LIGHT_THEME_COLOR_BOTTOM }}
          testID={'webview'} ref={webViewRef}
          source={{ uri: linkUrl }}
          cacheMode={'LOAD_NO_CACHE'}
          onMessage={handleMessage}
          onLoadEnd={() => {
            setInitialLoading(false);
          }}
          startInLoadingState={true}
          javaScriptEnabled={true}
          injectedJavaScript={injectedScript}
          onNavigationStateChange={handleNavState}
        />
      )}
    </SDKWrapperComponent>
  );
};
