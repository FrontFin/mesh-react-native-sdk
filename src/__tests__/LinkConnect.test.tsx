/* eslint-disable */
import React from 'react';
import { render } from '@testing-library/react-native';
import LinkConnect from '../components/LinkConnect';

jest.mock('react-native-webview', () => {
  const { View } = require('react-native');
  return {
    WebView: View,
  };
});

// Mock callback functions
const mockOnExit = jest.fn();
const SAMPLE_LINK_TOKEN = 'aHR0cHM6Ly93ZWIuZ2V0ZnJvbnQuY29tL2IyYi1pZnJhbWUvdGVzdC1hY2NvdW50LXJhbmRvbS9icm9rZXItY29ubmVjdC9jYXRhbG9nMQ==';

describe('LinkConnect Component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly when linkToken and accessTokens and transferDestinationTokens are provided', () => {
    render(<LinkConnect linkToken={SAMPLE_LINK_TOKEN} settings={{
      accessTokens: [
        {
          accountId: '1234567890',
          accountName: 'Test Account',
          accessToken: '1234567890',
          brokerType: 'test',
          brokerName: 'Test Broker',
        },
      ],
      transferDestinationTokens: [
        {
          accountId: '1234567890',
          accountName: 'Test Account',
          accessToken: '1234567890',
          brokerType: 'test',
          brokerName: 'Test Broker',
        },
      ],
    }} />);
  });

  it('renders correctly when linkToken is provided', () => {
    const tree = render(<LinkConnect linkToken={SAMPLE_LINK_TOKEN} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles onError callback for invalid linkToken', () => {
    render(<LinkConnect linkToken={'invalid=='} onExit={mockOnExit} />);

    expect(mockOnExit).toHaveBeenCalledWith('Invalid link token provided');
  });
});
