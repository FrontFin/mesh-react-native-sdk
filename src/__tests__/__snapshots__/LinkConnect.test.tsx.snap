// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkConnect Component renders correctly when linkToken is provided 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
  testID="link-connect-component"
>
  <RCTSafeAreaView
    style={
      {
        "backgroundColor": "#F3F4F5",
        "flex": 0,
      }
    }
  />
  <RCTSafeAreaView
    style={
      {
        "backgroundColor": "#FBFBFB",
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "#FBFBFB",
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 10,
          }
        }
      />
      <View
        bounces={false}
        cacheMode="LOAD_NO_CACHE"
        injectedJavaScript="
    window.meshSdkPlatform='reactNative';
    window.meshSdkVersion='2.1.19';
  "
        javaScriptEnabled={true}
        onLoadEnd={[Function]}
        onMessage={[Function]}
        onNavigationStateChange={[Function]}
        onShouldStartLoadWithRequest={[Function]}
        originWhitelist={
          [
            "*.meshconnect.com",
            "*.getfront.com",
            "*.walletconnect.com",
            "*.walletconnect.org",
            "*.walletlink.org",
            "*.coinbase.com",
            "*.okx.com",
            "*.gemini.com",
            "*.coinbase.com",
            "*.hcaptcha.com",
            "*.robinhood.com",
            "*.google.com",
            "https://meshconnect.com",
            "https://getfront.com",
            "https://walletconnect.com",
            "https://walletconnect.org",
            "https://walletlink.org",
            "https://coinbase.com",
            "https://okx.com",
            "https://gemini.com",
            "https://coinbase.com",
            "https://hcaptcha.com",
            "https://robinhood.com",
            "https://google.com",
            "https://front-web-platform-dev",
            "https://front-b2b-api-test.azurewebsites.net",
            "https://web.getfront.com",
            "https://web.meshconnect.com",
            "https://applink.robinhood.com",
            "https://m.stripe.network",
            "https://js.stripe.com",
            "https://app.usercentrics.eu",
            "robinhood://",
            "https://ramp.revolut.codes",
            "https://sso.revolut.codes",
            "https://ramp.revolut.com",
            "https://api.cb-device-intelligence.com",
          ]
        }
        source={
          {
            "uri": "https://web.getfront.com/b2b-iframe/test-account-random/broker-connect/catalog1?lng=en",
          }
        }
        startInLoadingState={true}
        style={
          {
            "backgroundColor": "#FBFBFB",
          }
        }
        testID="webview"
      />
    </View>
  </RCTSafeAreaView>
  <RCTSafeAreaView
    style={
      {
        "backgroundColor": "#FBFBFB",
        "flex": 0,
      }
    }
  />
</View>
`;
